app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/Model/ProductRepository/CleanGetProductIdsBySkusCache.php

```diff
@@ -0,0 +1,68 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalog\Plugin\CatalogInventory\Model\ProductRepository;
+
+use Magento\Catalog\Api\Data\ProductInterface;
+use Magento\Catalog\Model\ProductRepository;
+use Magento\InventoryCatalog\Model\LocalCache\GetProductIdsBySkusCache;
+
+/**
+ * Clean local cache on service GetProductIdsBySkus when product saved or deleted.
+ */
+class CleanGetProductIdsBySkusCache
+{
+    /**
+     * @var GetProductIdsBySkusCache
+     */
+    private $getProductIdsBySkusCache;
+
+    /**
+     * @param GetProductIdsBySkusCache $getProductIdsBySkusCache
+     */
+    public function __construct(
+        GetProductIdsBySkusCache $getProductIdsBySkusCache
+    ) {
+        $this->getProductIdsBySkusCache = $getProductIdsBySkusCache;
+    }
+
+    /**
+     * @param ProductRepository $subject
+     * @param ProductInterface $result
+     * @param ProductInterface $product
+     * @param bool $saveOptions
+     * @return ProductInterface
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function afterSave(
+        ProductRepository $subject,
+        ProductInterface $result,
+        ProductInterface $product,
+        $saveOptions = false
+    ): ProductInterface {
+        $this->getProductIdsBySkusCache->clean();
```

Comment:
the state is always bad. 
And if you provide an ability to explicitly manipulate with internal state - it's even worse

this is the root cause of the most unexpected bugs.

if needed you just need to vanish the whole object being used, but not to provide API for its internal state modification